cmd/pulumi-converter-terraform/par.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+)
+
+// Transforms a large map in batches of up to batch elements, using workers number of goroutines. If
+// workers is -1, uses one worker per CPU.
+func parTransformMap[K comparable, T any, U any](
+	inputs map[K]T,
+	transform func(map[K]T) (map[K]U, error),
+	workers int,
+	batch int,
+) (map[K]U, error) {
+	if batch < 1 {
+		return nil, fmt.Errorf("batch cannot be less than 1")
+	}
+	n := workers
+	if workers < 1 {
+		n = runtime.NumCPU()
+		if n < 2 {
+			n = 2
+		}
+	}
+
+	keys := []K{}
```

Comment:
I don't think you need a slice here. Just have three+ goroutines spinning.
input) a goroutine iterating the input map and feeding a channel with (key, value) pairs
N workers) reading from that channel and running the mapping function and writing to another (key, value) channel
output) reading from the channel workers is writing to and building a new map off those pairs

No temporary maps, no slice counting, and full use of all worker till everything is translated.

Reply:
Eh, sure.